### ISSUE TITLE
Decide on order of arguments for opening files

### ISSUE THREAD

**Issue by mhvk** [mhvk/baseband#112]
Ideally, order is by relevance. For writing, the overall logic seems obvious:
1. `header0` - ideally this has everything one needs;
2. Keywords not normally inferable from a header in a given format;
3. Keywords that define a full header.

For reading, one would only need (2), so the real question is about order of those. Going by what is essential in order of how the data is presented makes some sense, ie.,
1. `sample_rate`
2. organization over/in payload: `samples_per_frame`, `sample_shape` (or `nchan`, whatever defines it)
3. sample definition: `bps`, `complex_data`

Anyway, for discussion. cc @cczhu.

**Comment by mhvk** [comment]
* Mark4:
- needs `decade`
- `sample_rate` for single frame or writing
* Mark5b:
- needs `kday`, `nchan`; really `bps`
- `sample_rate` for less than 1 second or writing
* VDIF
- for some EDV, `sample_rate` for less than 1 second or writing
* DADA
- nothing

**Comment by cczhu** [comment]
Stream writers also have `header` for passing a prefab header.  I'd prefer it go just before `**kwargs`, though maybe not before `squeeze`?

**Comment by mhvk** [comment]
Ideally, the pre-fab header contains everything the writer needs, which would argue for putting it first... (and of course the reader always has that information, and determines some of the parameters from it).

**Comment by mhvk** [comment]
From #116, a concrete (if slightly out-of-date) proposal:

`open` parameters when reading (each format may have specific defaults or features that are additionally described):
```
    fh_raw : 
        File handle of the raw <FORMAT> stream
    sample_rate : `~astropy.units.Quantity`
        Number of complete samples per second (ie. the sampling rate of each <COMPONENTS>).
    ref_time : `~astropy.time.Time`, or None, optional
        Reference time within <TIME FRAME> of the start time of the observations.
    samples_per_frame : int
        Number of complete samples per frame.
    nthread : int
        Number of threads in a complete sample.
    nchan : int
        Number of channels in a complete sample.
    complex_data : bool
        Whether the data is complex.
    bps : int
        Bits per elementary sample (e.g., the real or imaginary part of one
        component of the complete sample).
    subset: indexing object, or tuple of objects
        Specific components of the complete sample to decode.  <SOME FORMAT SPECIFIC INSTRUCTIONS.>
    squeeze : bool
        If `True` (default), remove any dimensions of length unity from
        decoded data.
```

`open` parameters when writing:
```
    raw :
        Which will write filled sets of frames to storage.
    header : :class:`~baseband.vdif.VDIFHeader`
        Header for the first frame, holding time information, etc.
    squeeze : bool
        If `True` (default), ``write`` accepts squeezed arrays as input,
        and adds channel and thread dimensions if they have length unity.
    **kwargs
        If no header is given, an attempt is made to construct the header from
        these.  For a standard header, the following suffices.
```

**Comment by cczhu** [comment]
Here's my proposal for argument ordering across the formats.

1. header0 (switches spots with sample_rate when required)
2. sample_rate
3. organization over/in payload (including timestamp)
4. sample definition
5. user-defined payload slicing (reader only)
6. valid/verify (frame initializers)
7. kwargs

Annoyingly 1 switches with 2 for writers.  The alternative is to put header0 right before kwargs, but then it comes after verify, which for, say `Mark5BFileWriter.write_frame`, prevents the header and data from being grouped together.

**vlbi_base**:

- VLBIStreamBase: fh_raw, header0, sample_rate, samples_per_frame, unsliced_shape, bps, complex_data, subset, squeeze
- VLBIStreamReaderBase: fh_raw, header0, sample_rate, samples_per_frame, unsliced_shape, bps, complex_data, subset, squeeze

Both squeeze and sample_rate no longer have defaults for the vlbi classes.  That shouldn't be a problem since they're always passed by a child class.

**mark4**:

- Mark4StreamReader: fh_raw, sample_rate=None, ntrack=None, decade=None, ref_time=None, subset=None, sample_rate=None, squeeze=True
- Mark4StreamWriter: raw, sample_rate, header=None, squeeze=True, **kwargs

No changes needed to file reader/writer, payload/header fromfile/fromdata.

**mark5b**:

- Mark5bStreamReader: fh_raw, sample_rate=None, kday=None, ref_time=None, nchan=1, bps=2, subset=None, squeeze=True
- Mark5BStreamWriter: raw, header=None, sample_rate, nchan=1, bps=2, squeeze=True, **kwargs
- Mark5BFileWriter.write_frame: data, header=None, bps=2, valid=True, **kwargs
- Mark5BFrame.fromfile: fh, kday=None, ref_time=None, nchan=1, bps=2, valid=None, verify=True
- Mark5BFrame.fromdata: data, header=None, bps=2, valid=True, verify=True, **kwargs

For readers, nchan now defaults to 1, which is what the payload does eventually anyway.  I can also make it default to None which will throw an error (maybe do that for bps too?).

**vdif**:

- VDIFStreamBase: fh_raw, header0, sample_rate=None, nthread=1, subset=None, squeeze=True
- VDIFStreamReader: fh_raw, sample_rate=None, subset=None, squeeze=True
- VDIFStreamWriter: raw, header=None, sample_rate=None, nthread=1, squeeze=True, **kwargs
- VDIFFileReader.read_frameset: edv=None, thread_ids=None, sort=True, verify=True
- VDIFFrameSet.fromfile: fh, edv=None, thread_ids=None, sort=True, verify=True

**dada**:

Everything's fine!

**gsb**:

- GSBStreamBase: fh_ts, fh_raw, header0, sample_rate=None, samples_per_frame=None, payloadsize=None, nchan=None, bps=None, complex_data=None, subset=None, squeeze=True
- GSBStreamReader: fh_ts, fh_raw, sample_rate=None, samples_per_frame=None, payloadsize=None, nchan=None, bps=None, complex_data=None, subset=None, squeeze=True
- GSBStreamWriter: fh_ts, fh_raw, header=None, sample_rate=None, samples_per_frame=None, payloadsize=None, nchan=None, bps=None, complex_data=None, squeeze=True, **kwargs
- GSBPayload.fromfile: fh, payloadsize=None, nchan=1, bps=4, complex_data=False

**Comment by mhvk** [comment]
@cczhu - I like the ordering. I wonder if we shouldn't just remove even the switches. The only reason you have them is that for some writers the sample rate is required, while the header never is. But the only downside to just putting that as `(fh, header=None, sample_rate=None, ...)` is that we have to check that it is not actually `None` -- I think that is a small price for having the same order for all (and we need to do a sanity check on sample rate anyway...). Also, in the end, for the user it doesn't matter - people should really pass in by keyword.

Given this all-keyword logic, I think it makes some sense to have sensible defaults; `nchan=1` is arguably the only logical one, since anything else is just what we happen to have. I do think `bps=2` as a default for all the VLBI formats makes sense.

**Comment by mhvk** [comment]
I don't think we need more clarification, so removed that label!

**Comment by cczhu** [comment]
I agree, and will use `sample_rate=None` for writers.

For the case of GSB, where there's a default sample rate, should we force users to pass a sample rate to writers regardless?

**Comment by mhvk** [comment]
Fine with `GSB` having the default.

**Comment by mhvk** [comment]
fixed by #159
